src/vcpkg/sourceparagraph.cpp

```diff
@@ -1103,18 +1103,12 @@ namespace vcpkg
             return t;
         }
 
-        virtual void parse_name_version(vcpkg::Json::Reader& r,
-                                        const vcpkg::Json::Object& obj,
-                                        vcpkg::SourceParagraph& spgh) = 0;
-
-        virtual Optional<std::unique_ptr<SourceControlFile>> visit_object(Json::Reader& r,
-                                                                          const Json::Object& obj) override
+        vcpkg::Optional<std::unique_ptr<vcpkg::SourceControlFile>> visit_object_common(
```

Comment:
I'm not thrilled about this returning the `Optional<>`; I think it's very confusing that the function has a bunch of side effects and then _always_ just moves its last argument into the return.

I think it would be clearer at the call site to have
```cpp
visit_object_common(obj, spgh, r, *control_file);
return std::move(control_file);
```
versus
```cpp
return visit_object_common(obj, spgh, r, control_file);
```